File: web/src/audio/confidenceGate.ts
import { ManualMode } from "@state/useAppStore";

export interface GateConfig {
  lookAheadMs: number;
  attackMs: number;
  releaseMs: number;
  holdMs: number;
  thresholdOn: number;
  thresholdOff: number;
  framesOn: number;
  framesOff: number;
  duckDb: number;
}

const ZERO_DB = 0;

export class ConfidenceGate {
  private config: GateConfig;
  private sampleRate = 48000;
  private blockSize = 128;
  private gainDb = ZERO_DB;
  private targetDb = ZERO_DB;
  private holdTimerMs = 0;
  private consecutiveOn = 0;
  private consecutiveOff = 0;
  private manualMode: ManualMode = "auto";

  constructor(config: GateConfig) {
    this.config = config;
    this.reset();
  }

  configure(sampleRate: number, blockSize: number, config: GateConfig) {
    this.sampleRate = sampleRate;
    this.blockSize = blockSize;
    this.config = config;
    this.reset();
  }

  setManualMode(mode: ManualMode) {
    this.manualMode = mode;
  }

  reset() {
    this.gainDb = this.config.duckDb;
    this.targetDb = this.config.duckDb;
    this.holdTimerMs = 0;
    this.consecutiveOn = 0;
    this.consecutiveOff = 0;
  }

  update(confidence: number) {
    if (this.manualMode === "always_on") {
      this.targetDb = ZERO_DB;
    } else if (this.manualMode === "always_off") {
      this.targetDb = this.config.duckDb;
    } else {
      if (confidence >= this.config.thresholdOn) {
        this.consecutiveOn += 1;
        this.consecutiveOff = 0;
      } else if (confidence <= this.config.thresholdOff) {
        this.consecutiveOff += 1;
        this.consecutiveOn = 0;
      } else {
        this.consecutiveOn = 0;
      }

      if (this.consecutiveOn >= this.config.framesOn) {
        this.targetDb = ZERO_DB;
        this.holdTimerMs = this.config.holdMs;
      } else if (this.consecutiveOff >= this.config.framesOff && this.holdTimerMs <= 0) {
        this.targetDb = this.config.duckDb;
      }
    }

    const elapsedMs = (this.blockSize / this.sampleRate) * 1000;
    if (this.holdTimerMs > 0) {
      this.holdTimerMs = Math.max(0, this.holdTimerMs - elapsedMs);
    }

    const attackCoef = Math.exp(-elapsedMs / Math.max(this.config.attackMs, 1));
    const releaseCoef = Math.exp(-elapsedMs / Math.max(this.config.releaseMs, 1));

    if (this.gainDb > this.targetDb) {
      this.gainDb = this.targetDb + (this.gainDb - this.targetDb) * attackCoef;
    } else {
      this.gainDb = this.targetDb + (this.gainDb - this.targetDb) * releaseCoef;
    }

    this.gainDb = Math.min(ZERO_DB, Math.max(this.config.duckDb, this.gainDb));
    return this.gainDb;
  }

  currentGain() {
    return this.gainDb;
  }

  currentGainLinear() {
    return dbToLinear(this.gainDb);
  }
}

export function dbToLinear(db: number) {
  return Math.pow(10, db / 20);
}

File: desktop/src/dsp/ConfidenceGate.cpp
#include "dsp/ConfidenceGate.h"

#include <algorithm>
#include <cmath>

namespace singwithme::dsp
{
namespace
{
constexpr float kZeroDb = 0.0f;
constexpr float kMuteDb = -80.0f;
} // namespace

void ConfidenceGate::configure(float sampleRate, size_t blockSize, GateConfig config)
{
    sampleRate_ = sampleRate;
    blockSize_ = blockSize;
    config_ = config;
    gainDb_ = config_.duckDb;
    targetDb_ = config_.duckDb;
    holdTimerMs_ = 0.0f;
    consecutiveOn_ = 0;
    consecutiveOff_ = 0;
}

void ConfidenceGate::setManualMode(ManualMode mode)
{
    manualMode_ = mode;
}

float ConfidenceGate::update(float confidence, float vad, float pitch)
{
    (void)vad;
    (void)pitch;

    if (manualMode_ == ManualMode::AlwaysOn)
    {
        targetDb_ = kZeroDb;
    }
    else if (manualMode_ == ManualMode::AlwaysOff)
    {
        targetDb_ = config_.duckDb;
    }
    else
    {
        if (confidence >= config_.thresholdOn)
        {
            consecutiveOn_++;
            consecutiveOff_ = 0;
        }
        else if (confidence <= config_.thresholdOff)
        {
            consecutiveOff_++;
            consecutiveOn_ = 0;
        }
        else
        {
            consecutiveOn_ = 0;
        }

        if (consecutiveOn_ >= config_.framesOn)
        {
            targetDb_ = kZeroDb;
            holdTimerMs_ = config_.holdMs;
        }
        else if (consecutiveOff_ >= config_.framesOff && holdTimerMs_ <= 0.0f)
        {
            targetDb_ = config_.duckDb;
        }
    }

    const float elapsedMs = static_cast<float>(blockSize_) / sampleRate_ * 1000.0f;
    if (holdTimerMs_ > 0.0f)
    {
        holdTimerMs_ = std::max(0.0f, holdTimerMs_ - elapsedMs);
    }

    const float attackCoef = std::exp(-elapsedMs / std::max(config_.attackMs, 1.0f));
    const float releaseCoef = std::exp(-elapsedMs / std::max(config_.releaseMs, 1.0f));

    if (gainDb_ > targetDb_)
    {
        gainDb_ = targetDb_ + (gainDb_ - targetDb_) * attackCoef;
    }
    else
    {
        gainDb_ = targetDb_ + (gainDb_ - targetDb_) * releaseCoef;
    }

    gainDb_ = std::clamp(gainDb_, config_.duckDb, kZeroDb);
    return gainDb_;
}
} // namespace singwithme::dsp
